### PR TITLE
Revamp FilterStep to allow nested conditions trees

### PR DESCRIPTION
**Pull request by bloodstorms**
It misses conversion fonction to fix (value/int, value/float, value/bool)
And the filter schema errors

Except that, the code can be reviewed

Before:
Capture d’écran 2020-03-12 à 15 21 09

After:
Capture d’écran 2020-03-12 à 15 05 25
Capture d’écran 2020-03-12 à 15 06 57

### CODE REVIEW

File: src/components/stepforms/ConditionsEditor/ConditionsGroup.vue [hunk 1]
```diff
@@ -194,6 +192,33 @@ export default class ConditionsGroup extends Vue {
...
+      const newConditionsTree = {
+        ...this.conditionsTree,
+        operator: '',
+      } as AbstractFilterTree;
+      this.emitUpdatedConditionTree(newConditionsTree);
+    }
+    return;
+  }
+
+  setOperatorIfNecessaryAndUpdateConditionTree(newConditionsTree: AbstractFilterTree) {
```
Comment (rhuille) [inline src/components/stepforms/ConditionsEditor/ConditionsGroup.vue]:
```suggestion
  updateConditionTree(newConditionsTree: AbstractFilterTree) {
```
Only `updateConditionTree` is better I think

File: src/components/stepforms/FilterStepForm.vue [hunk 2]
```diff
@@ -74,32 +78,26 @@ export default class FilterStepForm extends BaseStepForm<FilterStep> {
...
-      'filter-form--multiple-conditions': this.conditions.length > 1,
-    };
+  get conditionsTree() {
+    return buildConditionsEditorTree(this.editedStep.condition);
   }
 
   submit() {
+    // FIXME !!!!
+    // This "if" works only for comboAnd, not for comboOr and simpleCondition, and not recursive for groups.
     if (isFilterComboAnd(this.editedStep.condition)) {
```
Comment (rhuille) [inline src/components/stepforms/FilterStepForm.vue]: Is it possible to do the fix now ? Why not ?

File: src/components/stepforms/ConditionsEditor/ConditionsGroup.vue [hunk 3]
```diff
@@ -147,17 +139,21 @@ export default class ConditionsGroup extends Vue {
     const newConditionsTree = {
       ...this.conditionsTree,
       groups: newGroups,
-    };
-    this.emitUpdatedConditionTree(newConditionsTree);
+    } as AbstractFilterTree;
+    // When there is only one condition, the operator is an empty string.
+    // Set the operator to 'and' when a group is added.
```
Comment (rhuille) [inline src/components/stepforms/ConditionsEditor/ConditionsGroup.vue]: This comment can be removed as it is inside `setOperatorIfNecessaryAndUpdateConditionTree`

File: src/components/stepforms/ConditionsEditor/ConditionsGroup.vue [hunk 4]
```diff
@@ -194,6 +192,33 @@ export default class ConditionsGroup extends Vue {
...
+        operator: '',
+      } as AbstractFilterTree;
+      this.emitUpdatedConditionTree(newConditionsTree);
+    }
+    return;
+  }
+
+  setOperatorIfNecessaryAndUpdateConditionTree(newConditionsTree: AbstractFilterTree) {
+    // When there is only one condition and no group, the operator is an empty string (cf resetOperatorIfNecessary).
+    // Set the operator to 'and' when the condition is no longer alone.
```
Comment (rhuille) [inline src/components/stepforms/ConditionsEditor/ConditionsGroup.vue]:
```suggestion
    // Set the operator to 'and' when a group is added.
```

File: src/components/stepforms/FilterStepForm.vue (hunk 2 above)
Comment (bloodstorms) [reply src/components/stepforms/FilterStepForm.vue]: Yeah, I just ran out of time...

File: src/components/stepforms/ConditionsEditor/ConditionsGroup.vue (hunk 4 above)
Comment (bloodstorms) [reply src/components/stepforms/ConditionsEditor/ConditionsGroup.vue]: hmm, it's not necessary when a group is added, it's when a group or a condition is added, and only in the case it's alone

File: src/components/stepforms/ConditionsEditor/ConditionsGroup.vue (hunk 1 above)
Comment (bloodstorms) [reply src/components/stepforms/ConditionsEditor/ConditionsGroup.vue]: ok :)

File: tests/unit/filter-step-form.spec.ts [hunk 5]
```diff
@@ -169,19 +185,17 @@ describe('Filter Step Form', () => {
...
     });
     // we're editing an existing step with column `foo`, therefore even if the
     // column `bar` is selected in the interface, the step column should remain
     // `foo`.
-    expect(wrapper.vm.$data.editedStep.condition.and[0].column).toEqual('foo');
+    expect(wrapper.vm.$data.editedStep.condition.column).toEqual('foo');
   });
 
-  it('should convert input value to integer when the column data type is integer', () => {
+  it.skip('should convert input value to integer when the column data type is integer', () => {
```
[the remark this answers is not in the document]
Comment (bloodstorms) [reply tests/unit/filter-step-form.spec.ts]: Yes, but these tests need to be fixed

File: tests/unit/filter-step-form.spec.ts [hunk 6]
```diff
@@ -6,26 +6,26 @@ describe('Filter Step Form', () => {
...
-      data: {
-        editedStep: {
-          name: 'filter',
-          condition: { and: [] },
-        },
-      },
-      errors: [{ keyword: 'minItems', dataPath: '.condition.and' }],
-    },
-  ]);
+  // runner.testValidationErrors([
```
[the remark this answers is not in the document]
Comment (bloodstorms) [reply tests/unit/filter-step-form.spec.ts]: Yes, but this test needs to be fixed

File: src/components/ConditionsEditor/ConditionsEditor.vue [hunk 7]
```diff
@@ -25,6 +25,24 @@
...
+        conditions: [
+          { column: 'Section', operator: 'eq',  value: 'A'}
+          { column: 'City', operator: 'neq',  value: 'Paris'}
+        ],
+        groups: []
+        operator: 'or'
+      }
+    ],
+    operator: 'and'
+  }
```
Comment (bloodstorms) [inline src/components/ConditionsEditor/ConditionsEditor.vue]: 👍

File: src/components/stepforms/convert-filter-step-tree.ts [hunk 8]
```diff
@@ -19,6 +31,33 @@ export function isFilterCombo(
...
+	operator: 'and',
+	conditions: [{ column: 'toto', value: 'A', operator: 'eq' }],
+	groups: [
+		operator: 'or',
+		conditions: [
+      { column: 'tata', value: 'B', operator: 'ne' },
+      { column: 'titi', value: '2', operator: 'le' }
+    ],
+	]
+}
```
Comment (bloodstorms) [inline src/components/stepforms/convert-filter-step-tree.ts]:
```suggestion
{
	operator: 'and',
	conditions: [{ column: 'toto', value: 'A', operator: 'eq' }],
	groups: [
		operator: 'or',
		conditions: [
      { column: 'tata', value: 'B', operator: 'ne' },
      { column: 'titi', value: '2', operator: 'le' }
    ],
    groups: [],
	]
}
```

File: src/components/stepforms/convert-filter-step-tree.ts [hunk 9]
```diff
@@ -53,6 +92,32 @@ export function buildConditionsEditorTree(
...
```
Comment (bloodstorms) [inline src/components/stepforms/convert-filter-step-tree.ts]: There is always the `groups` key

```suggestion
{
	operator: 'and',
	conditions: [{ column: 'toto', value: 'A', operator: 'eq' }],
	groups: [
		operator: 'or',
		conditions: [
      { column: 'tata', value: 'B', operator: 'ne' },
      { column: 'titi', value: '2', operator: 'le' }
    ],
    groups: [],
	]
}
```

File: src/components/stepforms/convert-filter-step-tree.ts (hunk 8 above)
Comment (rhuille) [reply src/components/stepforms/convert-filter-step-tree.ts]: Thanks ! I corrected it

File: src/components/stepforms/convert-filter-step-tree.ts (hunk 9 above)
Comment (rhuille) [reply src/components/stepforms/convert-filter-step-tree.ts]: Thanks ! I corrected it

File: src/components/stepforms/widgets/FilterSimpleCondition.vue [hunk 10]
```diff
@@ -2,21 +2,18 @@
   <div class="filter-form-simple-condition__container">
     <div class="filter-form-simple-condition-column-input">
       <AutocompleteWidget
-        :id="`${dataPath.replace(/[^a-zA-Z0-9]/g, '')}-columnInput`"
-        v-model="value.column"
+        id="columnInput"
```
Comment (rhuille) [inline src/components/stepforms/widgets/FilterSimpleCondition.vue]: This id will not be unique... That why I suggested: ` :id="`${dataPath.replace(/[^a-zA-Z0-9]/g, '')}-columnInput`"` Why did you remove it ? :)

File: src/components/stepforms/widgets/FilterSimpleCondition.vue [hunk 11]
```diff
@@ -2,21 +2,18 @@
...
+        @input="setSelectedColumns({ column: editedValue.column })"
         placeholder="Column"
         :data-path="`${dataPath}.column`"
         :errors="errors"
       />
     </div>
     <div class="filter-form-simple-condition-operator-input">
       <AutocompleteWidget
-        :id="`${dataPath.replace(/[^a-zA-Z0-9]/g, '')}-filterOperator`"
+        id="filterOperator"
```
Comment (rhuille) [inline src/components/stepforms/widgets/FilterSimpleCondition.vue]: Same feedback here

File: src/components/ConditionsEditor/ConditionsGroup.vue [hunk 12]
```diff
@@ -140,24 +143,28 @@ export default class ConditionsGroup extends Vue {
...
     });
 
     const newConditionsTree = {
       ...this.conditionsTree,
       groups: newGroups,
-    };
-    this.emitUpdatedConditionTree(newConditionsTree);
+    } as AbstractFilterTree;
+    // When there is only one condition, the operator is an empty string.
+    // Set the operator to 'and' when a group is added.
```
Comment (rhuille) [inline src/components/ConditionsEditor/ConditionsGroup.vue]:
```suggestion
```

File: src/components/ConditionsEditor/ConditionsGroup.vue [hunk 13]
```diff
@@ -140,24 +143,28 @@ export default class ConditionsGroup extends Vue {
...
       groups: [],
     });
 
     const newConditionsTree = {
       ...this.conditionsTree,
       groups: newGroups,
-    };
-    this.emitUpdatedConditionTree(newConditionsTree);
+    } as AbstractFilterTree;
+    // When there is only one condition, the operator is an empty string.
```
Comment (rhuille) [inline src/components/ConditionsEditor/ConditionsGroup.vue]:
```suggestion
```

File: src/components/ConditionsEditor/ConditionsGroup.vue (hunk 12 above)
Comment (rhuille) [reply src/components/ConditionsEditor/ConditionsGroup.vue]: We can remove this comment :) There are already in `setOperatorIfNecessaryAndUpdateConditionTree `
